src/flow/ui/FlowStackNavigator.tsx

```diff
@@ -1,18 +1,50 @@
 import React from 'react';
 
-import { FLOW_STACK_SCREEN_ROUTES, FLOW_STACK_SCREENS, INITIAL_FLOW_STACK_SCREEN_ROUTE } from '../nav/FlowStackRoutes';
+import { getInitialRouteNameStateContext } from '../../navigation/initialRouteNameState/InitialRouteNameStateContext';
+import useInitialRouteNameStateContext from '../../navigation/initialRouteNameState/useInitialRouteNameStateContext';
+
+import {
+  FLOW_STACK_SCREENS,
+  FlowStackRoute,
+  getFlowStackScreenRoutes,
+  routeIsFlowStackInitialRouteName,
+  routeIsFlowStackPreInitialRouteName,
+} from '../nav/FlowStackRoutes';
 import FlowStack from './FlowStack';
 
+const InitialRouteNameStateContext = getInitialRouteNameStateContext<FlowStackRoute>();
+
 const FlowStackNavigator = () => {
+  const { prevInitialRouteName, initialRouteName } = useInitialRouteNameStateContext(InitialRouteNameStateContext);
+
+  // Every time the initial route name changes to a different string, reconstruct the entire thing.
+  // Why? The initial route name will only be set once(!), i.e. during the initial render.
+  // Subsequent changes won't alter the initial route.
   return (
-    <FlowStack.Navigator initialRouteName={INITIAL_FLOW_STACK_SCREEN_ROUTE}>
-      {FLOW_STACK_SCREEN_ROUTES.map((route) => {
+    <FlowStack.Navigator>
+      {getFlowStackScreenRoutes(initialRouteName).map((route) => {
```

Comment:
[R] we might want to think about using a component state on this instead of relying a re-computation outside of the component. It is less apparent that the result of `getFlowStackScreenRoutes(initialRouteName)` can change upon different render cycle